Migrate etcd-quorum-guard from MCO.

retroflexer opened the pull request:
Migrating the quorum guard from machine-config-operator. 

Note that the TLS certs come from secrets and configmaps, instead of files from the host (MCO was using files on disk).

ironcladlou commented:
From an offline discussion worth bringing out here:

Instead of retaining CVO management of QG, this is an opportunity to bring the component under direct management of the operator. A static CVO-managed QG is simple but doesn't leave any room for us to enhance the component with dynamic behavior going forward. Is it low effort to bring QG under operator management?

retroflexer commented:
> From an offline discussion worth bringing out here:
> 
> Instead of retaining CVO management of QG, this is an opportunity to bring the component under direct management of the operator. A static CVO-managed QG is simple but doesn't leave any room for us to enhance the component with dynamic behavior going forward. Is it low effort to bring QG under operator management?

As we discussed on slack, I too think it is worthy idea to pursue.

But I would let this PR go through so  that the migration from MCO happens without glitches, and then, once we have the quorum guard in the etcd-operator namespace, we can move it under direct management of the operator.

retroflexer commented:
/retest

retroflexer commented:
/retest

ironcladlou commented:
>But I would let this PR go through so that the migration from MCO happens without glitches, and then, once we have the quorum guard in the etcd-operator namespace, we can move it the under direct management of the operator.

As long as we're committed to doing the operator transition within the same release, I guess I don't see the harm

retroflexer commented:
/retest

hexfusion commented:
> As long as we're committed to doing the operator transition within the same release, I guess I don't see the harm

My understanding is that moving a CVO managed resource is easy but removing is not. If we are really going to use the operator why remove 2 CVO resources?

hexfusion commented:
/hold 

I would like a detailed plan to address this.

vrutkovs commented on manifests/0000_12_etcd-operator_08_etcdquorumguard_deployment.yaml:
```diff
@@ -0,0 +1,102 @@
+  annotations:
+    exclude.release.openshift.io/internal-openshift-hosted: "true"
+spec:
+  replicas: 3
```
Could the operator control the amount of replicas? In OKD's fork we used https://github.com/openshift/machine-config-operator/commit/ddc89ca7814c38c88cfc0c27d083f82f899b5b85 in MCO - when `useUnsupportedUnsafeNonHANonProductionUnstableEtcd` is enabled etcd-quorum-guard is scaled back to 1

wking replied:
Operator controlling this Deployment so it can do intelligent things makes sense to me.  But can that happen after the pivot from machine-config operator to etcd operator?  We certainly don't want the etcd operator trying to do intelligent Deployment management while the machine-config operator is still trying to hand this Deployment off to the cluster-version operator, because the CVO and etcd operator would be fighting each other.

vrutkovs replied:
There should be any issue during transition from MCO to CEO as these are different deployments in different namespaces

wking replied:
> There should be any issue during transition from MCO to CEO as these are different deployments in different namespaces

Oh.  Why is the quorum guard not living in the same namespace as etcd?  Regardless, I'd still like the cross-repo pivot to be as boring and conservative as possible.  That way, if we flub the transition to intelligent etcd-operator management, we can easily revert the change without having to involve the MCO.  I don't think it's wrong to attempt the bigger pivot in a single PR, if folks are feeling more aggressive.  But is there much upside to moving to etcd-operator management in a single PR?

vrutkovs replied:
>Why is the quorum guard not living in the same namespace as etcd?

That would make MCO watch events in different namespace (currently etcd-quorum-guard lives in MCO namespace). 

>  I don't think it's wrong to attempt the bigger pivot in a single PR, if folks are feeling more aggressive

Sure, I don't mind splitting replica control in a separate PR

retroflexer commented:
@hexfusion  said:
> My understanding is that moving a CVO managed resource is easy but removing is not. If we are really going to use the operator why remove 2 CVO resources?
@wking Could you help me here? Is removing PDB not very easy? I was hoping just removing the pdb file is enough to get rid of it. What other implications and impacts we need to consider?

wking commented:
openshift/machine-config-operator#1928 is dropping some operator-specific e2e tests around this.  Does the etcd operator have a similar operator-specific e2e suite?  Or some other mechanism for ensuring that the quorum guard is operator as expected?

wking commented:
> I was hoping just removing the pdb file is enough to get rid of it.

That's true for _new_ clusters.  But existing clusters will still have the resource in-cluster.  If you're just transferring ownership, that's fine, because there's no need to actually remove the in-cluster resource.  You'd want the hand-off to look something like:

1. CVO manages the resource.
2. New release, operator begins managing the resource, but with a fixed target to match the CVO.  That way the operator and CVO are both controlling the resource, but they always agree on the intended state.
3. Drop the manifest, so the CVO stops trying to manage the resource.  The operator can now do whatever it likes for intelligent management, including removing the resource entirely.

The point of the second step is that during an update, the CVO will begin reconciling the new manifests.  If you jumped straight from 1 to 3, there would be a window where the CVO had relinquished control but the etcd operator had not yet been bumped to code that picked up control.  If the manifests for the etcd operator deployment and the managed resource are close together in the update graph, maybe that's fine.  If they are far enough apart that a stuck update could leave the resource unmanaged for a significant length of time, you probably want 2's careful handoff.

kikisdeliveryservice commented:
> [openshift/machine-config-operator#1928](https://github.com/openshift/machine-config-operator/pull/1928) is dropping some operator-specific e2e tests around this. Does the etcd operator have a similar operator-specific e2e suite? Or some other mechanism for ensuring that the quorum guard is operator as expected?

Seconding that moving https://github.com/openshift/machine-config-operator/blob/master/test/e2e/etcdquorumguard_test.go is necessary and I think is necessary to have it run on this PR/change.

retroflexer commented:
> [openshift/machine-config-operator#1928](https://github.com/openshift/machine-config-operator/pull/1928) is dropping some operator-specific e2e tests around this. Does the etcd operator have a similar operator-specific e2e suite? Or some other mechanism for ensuring that the quorum guard is operator as expected?

 I will work on adding e2e tests on our side, once we moved to more intelligent control by operator.

retroflexer commented:
/retest

retroflexer commented:
/retest

kikisdeliveryservice commented on test/e2e/etcquorumguard_test.go:
```diff
@@ -0,0 +1,288 @@
+
+type podinfo map[string]podstatus
+
+// TestEtcdQuorumGuard tests the etcd Quorum Guard.  It assumes there
```
HOORAY!

retroflexer commented:
Update: I added e2e test framework and added a test for EtcdQuorumGuard. 

I hope we can proceed with this PR to finish cross-repo transfer.

@wking @kikisdeliveryservice

retroflexer commented:
@kikisdeliveryservice I need your help again. The verify-deps is failing. I ran `go mod tidy;go mod verify; go mod vendor` and it was clean.

```
go: extracting github.com/xlab/handysort v0.0.0-20150421192137-fb3537ed64a1
github.com/openshift/cluster-etcd-operator/test/e2e tested by
	github.com/openshift/cluster-etcd-operator/test/e2e.test imports
	github.com/openshift/cluster-etcd-operator/test/e2e/framework: no matching versions for query "latest"
```
I appreciate any help.

retroflexer commented:
/retest

hexfusion commented on go.mod:
```diff
@@ -6,10 +6,12 @@ require (
 	github.com/davecgh/go-spew v1.1.1
 	github.com/ghodss/yaml v1.0.0
 	github.com/go-bindata/go-bindata v3.1.2+incompatible
+	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
```
since we already have klog can we use that instead?

retroflexer replied:
Yes. Thanks. I got rid of glog.

kikisdeliveryservice commented:
> @kikisdeliveryservice I need your help again. The verify-deps is failing. I ran `go mod tidy;go mod verify; go mod vendor` and it was clean.
> 
> ```
> go: extracting github.com/xlab/handysort v0.0.0-20150421192137-fb3537ed64a1
> github.com/openshift/cluster-etcd-operator/test/e2e tested by
> 	github.com/openshift/cluster-etcd-operator/test/e2e.test imports
> 	github.com/openshift/cluster-etcd-operator/test/e2e/framework: no matching versions for query "latest"
> ```
> 
> I appreciate any help.

You import "github.com/openshift/cluster-etcd-operator/test/e2e/framework" in test/e2e/etcquorumguard_test.go but you don't actually have that dir and contents anywhere :)

retroflexer commented:
> You import "github.com/openshift/cluster-etcd-operator/test/e2e/framework" in test/e2e/etcquorumguard_test.go but you don't actually have that dir and contents anywhere :)

Oops. Looks like I missed checking in clientset.go in framework. Thanks. It was a long day. Phew.

hexfusion commented on Makefile:
```diff
@@ -35,3 +35,15 @@ $(call add-bindata,etcd,./bindata/etcd/...,bindata,etcd_assets,pkg/operator/etcd
+	CGO_ENABLED=0 go test -tags=$(GOTAGS) -count=1 -v ./cmd/... ./pkg/... 
+
+# This was copied from https://github.com/openshift/cluster-image-registry-operator
+test-e2e:
```
for e2e can we align with KAS?

https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/Makefile#L70

Also we already have `test-unit` right? 

https://github.com/openshift/release/blob/master/ci-operator/config/openshift/cluster-etcd-operator/openshift-cluster-etcd-operator-release-4.6.yaml#L51

retroflexer replied:
`test-unit` was failing with e2e added. See previous test outputs:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/394/pull-ci-openshift-cluster-etcd-operator-master-unit/1285873347153891328

retroflexer commented:
/test e2e-gcp-upgrade

retroflexer commented:
/retest

retroflexer commented:
/test e2e-gcp-upgrade

hexfusion commented:
/test e2e-gcp-upgrade

retroflexer commented:
/test e2e-gcp-upgrade

retroflexer commented:
All the required (at least the 'Required' ones) appear to be passing the tests. I am testing the y-upgrade to be working using cluster-bot (to make sure there will not be two quorum guards at the end of the upgrade).

@wking do you see any problems with going ahead with this PR?

hexfusion commented:
Lets step through next steps.

1.) merge this we now have 2 quorum-guard deployments.
2.) state a PR to MCO removing quorum-guard and PDB manifests.
3.) any other actions required to make sure that both Z and Y upgrades have only a single deployment?

wking commented on Makefile:
```diff
@@ -35,3 +35,15 @@ $(call add-bindata,etcd,./bindata/etcd/...,bindata,etcd_assets,pkg/operator/etcd
+
+# Unit tests only (no active cluster required)
+test-unit:
+	CGO_ENABLED=0 go test -tags=$(GOTAGS) -count=1 -v ./cmd/... ./pkg/... 
```
nit:

```console
$ git --no-pager show --oneline --check origin/pr/394
f611f340 (origin/pr/394) Add e2e test framework and add a test for EtcdQuorumGuard
Makefile:45: trailing whitespace.
+	CGO_ENABLED=0 go test -tags=$(GOTAGS) -count=1 -v ./cmd/... ./pkg/... 
```

hexfusion replied:
@retroflexer I brought this up but I am not sure why we must redefine test-unit. KAS has both unit and e2e tests.

hexfusion replied:
```
vendor/github.com/openshift/build-machinery-go/make/golang.mk
```

retroflexer replied:
@hexfusion We need to overwrite the default rules. The default test-unit is attempting to run e2e tests as well, and was failing. By specifying only ./cmd and ./pkg directories, we are excluding it from running the tests in ./e2e directory.

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/394/pull-ci-openshift-cluster-etcd-operator-master-unit/1285873347153891328

retroflexer replied:
I deleted these overriding rules, and followed the examples from  the Makefile of https://github.com/openshift/cluster-kube-apiserver-operator. Now we have to wait for the all the tests to pass again :-).

wking commented on Makefile:
```diff
@@ -35,3 +35,15 @@ $(call add-bindata,etcd,./bindata/etcd/...,bindata,etcd_assets,pkg/operator/etcd
+
+# This was copied from https://github.com/openshift/cluster-image-registry-operator
+test-e2e:
+	go test -failfast -timeout 120m -v$${WHAT:+ -run="$$WHAT"} ./test/e2e/
```
is `WHAT` a think here?  If so, did you want to land docs similar to what origin [used to have][1]?  Or somewhere explain to folks how it is expected to be used?

[1]: https://github.com/openshift/origin/commit/82d3e63a7e267292a28547b51bc0b79f2c7e2213#diff-b67911656ef5d18c4ae36cb6741b7965L147

retroflexer replied:
I can removed. I think `go test` allows a regex pattern to be specified to only run a subset of tests. I can remove it, if you like.

```
-run regexp
        Run only those tests and examples matching the regular expression.
        For tests, the regular expression is split by unbracketed slash (/)
        characters into a sequence of regular expressions, and each part
        of a test's identifier must match the corresponding element in
        the sequence, if any. Note that possible parents of matches are
        run too, so that -run=X/Y matches and runs and reports the result
        of all tests matching X, even those without sub-tests matching Y,
        because it must run them to look for those sub-tests.
```

retroflexer replied:
I deleted the overriding rules for test-e2e and test-unit and replaced them with definitions similar to what the Makefile from https://github.com/openshift/cluster-kube-apiserver-operator has. Let's wait for the tests to succeed again.

retroflexer commented:
/retest

retroflexer commented:
/retest

hexfusion commented:
/test e2e-gcp-upgrade

retroflexer commented:
/test e2e-gcp-upgrade

retroflexer commented:
/test e2e-gcp-upgrade

retroflexer commented:
/retest

retroflexer commented:
/retest

hexfusion commented:
/hold cancel

hexfusion commented:
@retroflexer has manually tested upgrades with success.
/approve

retroflexer commented:
Verified the PR to be working well with y-upgrades using cluster-bot on aws and gcp (tested with MCO's 1928 along with this PR).

aws:
job test upgrade 4.5 4.6,openshift/machine-config-operator#1928,openshift/cluster-etcd-operator#394 aws succeeded
(https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1286728540858880000)

gcp:
job test upgrade 4.5 4.6,openshift/machine-config-operator#1928,openshift/cluster-etcd-operator#394 gcp succeeded
(https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-gcp/1286844014271664128)

wking commented:
[disruptive][1]:

```
Jul 25 07:41:36.324: INFO: Running AfterSuite actions on node 1
fail [github.com/openshift/origin@/test/extended/dr/quorum_restore.go:195]: Unexpected error:
    <*errors.StatusError | 0xc0023a0e60>: {
        ErrStatus: {
            TypeMeta: {Kind: "Status", APIVersion: "v1"},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "rpc error: code = Unknown desc = context deadline exceeded",
            Reason: "",
            Details: nil,
            Code: 500,
        },
    }
    rpc error: code = Unknown desc = context deadline exceeded
occurred

failed: (38m17s) 2020-07-25T07:41:36 "[sig-etcd][Feature:DisasterRecovery][Disruptive] [Feature:EtcdRecovery] Cluster should restore itself after quorum loss [Disabled:Broken] [Serial] [Suite:openshift]"
```

Any idea what's going on there?

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/394/pull-ci-openshift-cluster-etcd-operator-master-e2e-aws-disruptive/1286912222311747584

retroflexer commented:
> fail [github.com/openshift/origin@/test/extended/dr/quorum_restore.go:195]: Unexpected error:
>    <*errors.StatusError | 0xc0023a0e60>: {

@wking This quorum restore is regarding the restoration of the backup and has nothing to do with quorum guard per se. These tests haven't passed in a while, and I agree we need to fix these tests to make them work reliably. But none of the changes in this PR, I believe, are impacting the results.

hexfusion commented:
/lgtm

merging with the understanding that having 2 quorum-guards for a short period of time will not cause failure.

retroflexer commented:
/retest

retroflexer commented:
/test e2e-gcp-upgrade
